Components/Helper.php

```diff
@@ -924,6 +907,26 @@ public function addShopIdToConnectCategories($offset, $batchsize)
     }
 
     /**
+     * @param $categoryKey
+     * @param $shopId
+     * @param $category
+     * @param $row
+     */
+    private function addShopIdToConnectCategory($categoryKey, $shopId, $category, $row)
+    {
+        $selectedCategory = $this->manager->getConnection()->executeQuery('SELECT `id` FROM s_plugin_connect_categories WHERE category_key = ? AND shop_id = ?',
+            [$categoryKey, $shopId]);
+        if (!($res = $selectedCategory->fetch())) {
+            $categoryId = $this->createCategoryWithShopId($categoryKey, $shopId, $category);
+        } else {
+            $categoryId = (int)$res['id'];
+        }
+        $this->manager->getConnection()->executeQuery('INSERT INTO s_plugin_connect_product_to_categories (articleID, connect_category_id) VALUES (?, ?)',
+            [$row['article_id'], $categoryId]
+        );
```

Comment:
I meant to move all the logic from this private method to ```createCategoryWithShopId```
Then you can use ``` $selectedCategory = $this->manager->getConnection()->executeQuery('SELECT `id` FROM s_plugin_connect_categories WHERE category_key = ? AND shop_id IS NULL',
            [$categoryKey]);```  only once. Don't need to fetch it by category key and shop id. If it doesn't exist then call ```$this->createCategoryWithShopId($categoryKey, $shopId, $category);```.

My idea is to call the query above then if category doesn't exist or it has different shopID insert new record in s_plugin_connect_categories

Reply:
But i have to fetch this by shopId and categoryKey to know if it already exists
it's important that i fetch one time with the exact shopId and one time sourceId should be null
that are two different cases i can't do it in one query